doc/cpp-style-guide.md

```diff
@@ -0,0 +1,933 @@
+* [protobuf](https://github.com/google/protobuf): this is a hard dependencies for gRPC.
+* [googletest](https://github.com/google/googletest): this is a requirement for gRPC and protobuf, and used only for
+    testing.
+* The C++11 standard library.
+
+This is a fairly minimal set. Any proposal to add new libraries will require careful consideration. Additional libraries
+add complexity to our users. Each additional library increases the chances that there will be an incompatible version
+installed in their system or already part of their application.
+
+## Guidelines
+
+These guidelines are applicable to all the C++ code in `google-cloud-cpp`.  When this guideline matches the GSG, we
+only describe the decisions in this document, and link to the GSG for rationale and examples.
+
+### Header Files
+
+In general, every `.cc` file should have an associated `.h` file. There are some common exceptions, such as unittests
+and small `.cc` files containing just a `main()` function.
+
+Correct use of header files can make a huge difference to the readability, size and performance of your code.
+
+The following rules will guide you through the various pitfalls of using header files.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers)
+
+#### Self-contained Headers
+
+Header files should be self-contained (compile on their own) and end in `.h`. Non-header files that are meant for
+inclusion should end in `.inc` and be used sparingly.
+
+All header files should be self-contained. Users and refactoring tools should not have to adhere to special
+conditions to include the header. Specifically, a header should have header guards and include all other headers it
+needs.
+
+Prefer placing the definitions for template and inline functions in the same file as their declarations. The
+definitions of these constructs must be included into every `.cc` file that uses them, or the program may fail to link
+in some build configurations. If declarations and definitions are in different files, including the former should
+transitively include the latter. Do not move these definitions to separately included header files (`-inl.h`); this
+practice was common in the past, but is no longer allowed.
+
+As an exception, a template that is explicitly instantiated for all relevant sets of template arguments, or that is a
+private implementation detail of a class, is allowed to be defined in the one and only `.cc` file that instantiates
+the template.
+
+There are rare cases where a file designed to be included is not self-contained. These are typically intended to be
+included at unusual locations, such as the middle of another file. They might not use header guards, and might not
+include their prerequisites. Name such files with the `.inc` extension. Use sparingly, and prefer self-contained
+headers when possible.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers)
+
+#### The `#define` Guard
```

Comment:
Is use of #pragma once an acceptable substitute?

Reply:
In general, I tried to not deviate from the Google Style Guide too much.  `#pragma once` is widely supported, but it is still messy to figure out if it is, look at:

https://en.wikipedia.org/wiki/Pragma_once#Portability

I guess we could do:

```C++
#include "bigtable/check_pragma_once_support.h" // find a better name.
#if GOOGLE_CLOUD_CPP_HAVE_PRAGMA_ONCE
#  pragma once
#endif  // GOOGLE_CLOUD_CPP_HAVE_PRAGMA_ONCE
```

Reply:
Given the wide availability of pragma once on known compilers, maybe we can allow use without adding any checks and revisit if we receive reports of compilers that dont support it?